src/systems/track_controller/TrackController.cc

```diff
@@ -0,0 +1,579 @@
+
+  /// \brief The callback for CollectContactSurfaceProperties - all the magic
+  /// happens here.
+  /// \param[in] _collision1 The first colliding body.
+  /// \param[in] _collision2 The second colliding body.
+  /// \param[in] _point The contact point (in world coords).
+  /// \param[in] _force Force in the contact point (may be omitted).
+  /// \param[in] _normal Unit normal of the collision.
+  /// \param[in] _depth Depth of penetration.
+  /// \param[in] _numContactsOnCollision Number of contacts that share the same
+  ///            collision body.
+  /// \param[inout] _params The contact surface parameters to be set by this
+  ///               system.
+  public: void ComputeSurfaceProperties(
+    const Entity& _collision1,
+    const Entity& _collision2,
+    const Eigen::Vector3d & _point,
+    const Eigen::Vector3d * const _normal,
+    F::ContactSurfaceParams<P>& _params);
+
+  /// \brief Compute speed and direction of motion of the contact surface.
+  /// \param[in] _beltSpeed Speed of the belt.
+  /// \param[in] _beltDirection Direction of the belt (in world coords).
+  /// \param[in] _frictionDirection First friction direction (in world coords).
+  /// \return The computed contact surface speed.
+  public: double ComputeSurfaceMotion(
+    double _beltSpeed, const ignition::math::Vector3d &_beltDirection,
+    const ignition::math::Vector3d &_frictionDirection);
+
+  /// \brief Compute the first friction direction of the contact surface.
+  /// \param[in] _centerOfRotation The point around which the track circles (
+  ///                              +Inf vector in case of straight motion).
+  /// \param[in] _contactWorldPosition Position of the contact point.
+  /// \param[in] _contactNormal Normal of the contact surface (in world coords).
+  /// \param[in] _beltDirection Direction of the belt (in world coords).
+  public: ignition::math::Vector3d ComputeFrictionDirection(
+    const ignition::math::Vector3d &_centerOfRotation,
+    const ignition::math::Vector3d &_contactWorldPosition,
+    const ignition::math::Vector3d &_contactNormal,
+    const ignition::math::Vector3d &_beltDirection);
+
+  /// \brief Name of the link to which the track is attached.
+  public: std::string linkName;
+
+  /// \brief Orientation of the track relative to the link. It is assumed that
+  /// the track moves along the +x direction of the transformed coordinate
+  /// system.
+  public: math::Quaterniond trackOrientation;
+
+  /// \brief Enables debugging prints and visualizations.
+  public: bool debug {false};
+  /// \brief Cached marker message for debugging purposes.
+  public: msgs::Marker debugMarker;
+  /// \brief ID of the debug marker. Should reset to 0 at each iteration start.
+  public: uint64_t markerId;
+
+  /// \brief Event manager.
+  public: EventManager* eventManager;
+  /// \brief Connection to CollectContactSurfaceProperties event.
+  public: common::ConnectionPtr eventConnection;
```

Comment:
For some reason, ign-gazebo segfaults during destruction of `eventConnection`. I'm working on finding out why.

Reply:
https://github.com/ignitionrobotics/ign-gazebo/issues/906

Reply:
Thanks for investigating